implementations/rust/ockam/ockam_api/src/schema/schema.cddl

```diff
@@ -249,3 +249,48 @@ activate_subscription = {
     ?3: text,       ;; space_name
     ?4: [+ text]    ;; owner_emails
 }
+
+;;; base ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+node_status = {
+     1: text,       ;; node_name
+     2: text,       ;; status
+     3: uint,       ;; workers
+     4: int,        ;; pid
+}
+
+;;; credentials ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+get_credential_request = {
+     1: bool,       ;; overwrite
+    ?2: text,       ;; identity_name
+}
+
+present_credential_request = {
+     1: text,       ;; route
+     2: bool,       ;; oneway
+}
+
+;;; flow_controls ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+add_consumer = {
+     1: flow_control_id,    ;; id
+     2: multi_addr,         ;; address
+}
+
+flow_control_id = {
+    1: text                 ;; id
+}
+
+multi_addr = bytes
+
+;;; policy ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+policy_list = {
+    1: [* expression]                                   ;; expressions
+}
+
+expression = [action expr]                              ;; Expression
```

Comment:
This grammar does not follow exactly the struct definitions for `Expr`:

 - `action` is `(nil text)`. Why `nil`?
 - `1..7 expr_data` instead of `* expr_data`
 - no distinction between `Expr::Seq` and `Expr::List`
 - there is no `Expr::Ident`

I am not really familiar with CDDL and there might be good reasons why you did it this way. Don't you think we should spend more time to make the grammar more precise?

Reply:
#1 why nil: stupid argument: it won't work without it :). The nil is there in the binary cbor (246_u8 to be precise) . This may even be an bug in mini-cbor as far as I know. NB this expresssion is not a choice, it's a ccdl list, so nil followed by text is how the Action struct is serialized. Maybe this is just the way tuple structs are handled, because they don't have named fields. 

#2 I don't see [* expr_data]. You're referring to [+ expr_data]? I find that recursive definition quite accurate as it reflects the one in Expr

#3 in cbor representation there is only the enum constant (6 or 7) so that's how it's distinguished when (de)serializing. The enum data is the same. 

#4 It's there as enum constant 5. Contents are same as Str, (String -> text) so that's the reason it's not mentioned explicitly

Yes I suspect that the expr cddl can be somewhat more precise. Will look into that

Reply:
1. The cddl is now both clearer and more strict. Please have a look.|
2. I found the actual reason for Nil

This is the definition:
pub struct Expression {
    #[n(1)]
    action: Action,
    #[n(2)]
    expr: Expr,
}

which is encoded as a list. Because the list is 0-indexed, but the first attribute (Action) hs #[n(1)], so Nil is inserted at index 0 to put action at index 1 in the list. One way to fix this is to replace 1 and 2 by 0 and 1, so that Nil can just disappear from the CDDL and you save 1 byte in the serialized message. Love to hear your thoughts on that